### ISSUE TITLE
Safe to use custom Query subclasses?

### ISSUE THREAD

**Issue by benny-medflyt** [brianc/node-postgres#1389]
It is possible to pass `client.query` an instance of a `Query` object. This is not documented (but it can be seen in some of the tests).

How stable is the "Query" interface? I would like to override the `handleRowDescription` and `handleDataRow` functions, but I am worried that because none of this is documented (or has tests), future versions of "pg" might change the implementation of Query (or change the way Client and Query interact) which will cause breakage.

**Comment by abenhamdine** [comment]
Actually it's documented incidentally here : https://node-postgres.com/guides/upgrading#client-query-submittable-

**Comment by benny-medflyt** [comment]
Thanks, but those docs still appear to be incomplete. It says that the object must have a "submit" method. But from looking at the source code, as well as the source code of "pg-cursor", the object must also have these (undocumented) methods:

    - handleError
    - handleReadyForQuery
    - handleError
    - handleRowDescription
    - handleDataRow
    - handlePortalSuspended
    - handleEmptyQuery
    - handleCommandComplete
    - handleCopyInResponse (not implemented in pg-cursor)
    - handleCopyData (not implemented in pg-cursor)

But thanks for pointing me in the direction of pg-cursor. Its existence is good evidence that using these methods is probably supported

**Comment by brianc** [comment]
Yeah I need to flesh out the docs on Submittable more & figure out exactly what kind of extensibility can be supported.  That interface isn't entirely stable in that it's not documented, but it hasn't changed in *years* & I doubt it will in the future.  No _promises_ until I can do the documentation on it, but it's very unlikely to change.

**Comment by brianc** [comment]
note: leaving this open until I can finish the documentation that @abenhamdine pointed out. 😄

**Comment by benny-medflyt** [comment]
cool. For my use case it is important for me to have access to the raw rows that are returned from the database (string values or "null"). Ideally as they come in one-by-one.

Overriding Query class and overriding `handleDataRow` works perfectly 👍 
so hopefully this won't go away in future versions =]

**Comment by brianc** [comment]
@benny-medflyt - It wont go away in the future! I'm going to document the entire API at some point. It's not documented yet because I'm limited on time and it's a lower priority thing.
